proposals/0000-surface-async-stream-errors-to-http.md

```diff
@@ -0,0 +1,231 @@
+
+In the common case of the backend speaking HTTP/1.1 to a reverse proxy/front-end terminator/CDN/etc. that translates to HTTP/2 for browsers, what can be done?
+
+## Migration strategy
+
+> Is this proposal backwards incompatible? Does this proposal replace an existing feature/pattern? If so, can we safely and automatically migrate existing apps/components to this proposal? How?
+
+> If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+
+## Drawbacks
+
+So far, Marko is totally agnostic as to what kind of stream it’s outputting to. You could use it to stream to a file, or something really unexpected like an FTP stream transmission or a chat protocol.
+
+It’s possible the server runtime might expose a different method to avoid breaking such cases, like `template.streamToHttp()`.
+
+However, there’s already a lot about Marko’s server rendering that only makes sense delivered over HTTP. For example, if you were rendering to a file to be served later, you’d want to buffer the component init scripts and you’d never want `client-reorder` enabled.
+
+## Alternatives
+
+The impact of not doing this would be no change; we would continue exposing Marko authors to the risks mentioned in the Motivation section.
+
+In the past, browsers other than Internet Explorer accepted `multipart/x-mixed-replace` responses to be loaded in `target="_top"`, and that could potentially be used to swap to a a new copy of the response-in-progress but with error-indicating headers, or potentially even a replacement error page. However, modern browsers have quietly dropped support for `multipart/x-mixed-replace` in the top browsing context.
+
+It _might_ be possible to write `<meta http-equiv="refresh" content="0;url=${errorPageLocation}">`. It’s not _supposed_ to be outside the `<head>`, but I still like it better than how Rails creates a mid-stream error redirect with an inline `<script>`. ([some context on how search engines could better understand this is a temporary redirect, not a permanent one](https://twitter.com/JohnMu/status/969486943351394304))
+
+
+## Prior art
+
+[You used to have to delay a bit before closing a connection without the zero-length terminator because of this Chrome bug](https://bugs.chromium.org/p/chromium/issues/detail?id=610126)
+
+https://stackoverflow.com/questions/17203379/response-sent-in-chunked-transfer-encoding-and-indicating-errors-happening-after
+
+The issue of signaling errors that occur during a dynamic HTTP stream is not specific to Marko; the problem is very old. Other server runtimes and frameworks may have paved this cowpath already. Some candidates widely-used enough to probably have built-in behavior or guidance:
+
+- Express and other Node HTTP frameworks
+    - https://expressjs.com/en/guide/error-handling.html#the-default-error-handler
+    - https://github.com/expressjs/express/issues/2700 (!!!)
```

Comment:
This is interesting. If express's `next`/error handler _does_ support this properly then perhaps using the `marko/express` plugin should be good enough. We actually [forward Marko's errors to express in that case](https://github.com/marko-js/marko/blob/master/packages/marko/src/express.js#L56).

Reply:
Only uncaught errors though.  The case we need to think about is when the error was caught and so we were able to render a page for the user, but we want to signal that content is missing to any proxies/bots/etc. that might otherwise cache/index a page that didn't render all its content.

If we did want to rely on this for those cases, we'd probably have to hold off on firing the event until just before Marko ends the stream, otherwise express would close the connection before we were done rendering.  We'd also probably want to use an event other than `error` since I know there's already apps out there listening for that event and doing a `script`/`meta` redirect.